### DISCUSSION TITLE
Bc for multiple pde's

### DISCUSSION THREAD

**Discussion by s133314** [zwicker-group/py-pde#79] [Q&A] [answered]
I have a basic convection-diffusion-reaction system and got most of everything implemented based on the Brusselator example. However, I want to supply different boundary conditions for the two equations. (Equation is d<sub>t</sub>c<sub>i</sub>=-ud<sub>x</sub>c<sub>i</sub>+Dd<sub>xx</sub>c<sub>i</sub> - kc_i)
I tried a couple of things which were not succesfull. One of my approaches was:
```python
eq = pde.PDE({'c0': f"-{u[0]}*get_x(gradient(c0)) + {d[0]}*laplace(c0) - {k[0]}*c0",
              'c1': f"-{u[1]}*get_x(gradient(c1)) + {d[1]}*laplace(c1) - 2*{k[1]}*c1"},
             user_funcs={"get_x": lambda arr: arr[0]}, 
             bc=[{"value": bc_c0, "value": bc_c1}, #Two dirchlet bc at the left with different value 
                 {"derivative": 0.0}]) #Neuman bc for both c0 and c1 at the right
```
This approaches doesn't give any errors but all equations have the same left bc...
I also tried 
```python
eq = pde.PDE({'c0': f"-{u[0]}*get_x(gradient(c0)) + {d[0]}*laplace(c0) - {k[0]}*c0",
              'c1': f"-{u[1]}*get_x(gradient(c1)) + {d[1]}*laplace(c1) - 2*{k[1]}*c1"},
             user_funcs={"get_x": lambda arr: arr[0]}, 
             bc=[{"value": numpy.array([bc_c0, bc_c1])}, #Two dirchlet bc at the left with different value 
                 {"derivative": 0.0}]) #Neuman bc for both c0 and c1 at the right
```
But this gave an error that `name 'laplace'is not defined`

**Comment by s133314** [comment, marked as answer]
Aah I found the solution while reading another post and a bit more in the documentation. I shouldn't use bc but bc_ops.
The PDE would be defined as such than:
```python
eq = pde.PDE({'c0': f"-{u[0]}*get_x(gradient(c0)) + {d[0]}*laplace(c0) - {k[0]}*c0",
              'c1': f"-{u[1]}*get_x(gradient(c1)) + {d[1]}*laplace(c1) - 2*{k[1]}*c1"},
             user_funcs={"get_x": lambda arr: arr[0]}, 
             bc_ops={"c0:*": [{"value": bc_c0},  {"derivative": 0.0}],
                     "c1:*": [{"value": bc_c1},  {"derivative": 0.0}]})
````

**Reply by sashahexe** [reply]
> Aah I found the solution while reading another post and a bit more in the documentation. I shouldn't use bc but bc_ops. The PDE would be defined as such than:
> 
> ```python
> eq = pde.PDE({'c0': f"-{u[0]}*get_x(gradient(c0)) + {d[0]}*laplace(c0) - {k[0]}*c0",
>               'c1': f"-{u[1]}*get_x(gradient(c1)) + {d[1]}*laplace(c1) - 2*{k[1]}*c1"},
>              user_funcs={"get_x": lambda arr: arr[0]}, 
>              bc_ops={"c0:*": [{"value": bc_c0},  {"derivative": 0.0}],
>                      "c1:*": [{"value": bc_c1},  {"derivative": 0.0}]})
> ```

Hi! I am working on a similar problem. Could I ask you to share your code if you'd be so kind? I'm getting errors when trying to use this solution for different bcs. Thank you!

**Reply by s133314** [reply]
```python
    eq = pde.PDE({'c0': f"-{simulation_parameters['u'][0]}*get_x(gradient(c0)) + {simulation_parameters['D'][0]}*laplace(c0)",
                  'c1': f"-{simulation_parameters['u'][0]}*get_x(gradient(c1)) + {simulation_parameters['D'][1]}*laplace(c1) - {simulation_parameters['k']}*c1**1",
                  'c2': f"-{simulation_parameters['u'][0]}*get_x(gradient(c2)) + {simulation_parameters['D'][2]}*laplace(c2) + {simulation_parameters['k']}*c1**1"},
             user_funcs={"get_x": lambda arr: arr[0]}, 
             bc_ops={"c0:*": [{"value": simulation_parameters["bc"][0]}, {"derivative": 0.0}],
                     "c1:*": [{"value": simulation_parameters["bc"][1]}, {"derivative": 0.0}],  #Dirichlet one side, Neumann other side, not sure how it would work for 2D or 3D
                     "c2:*": [{"value": simulation_parameters["bc"][2]}, {"derivative": 0.0}]})

    # initialize state
    grid = pde.CartesianGrid([[0, 1]], nx, periodic=False)
    state = pde.FieldCollection([pde.ScalarField(grid, simulation_parameters["ini"][0]),
                                 pde.ScalarField(grid, simulation_parameters["ini"][1]),
                                 pde.ScalarField(grid, simulation_parameters["ini"][2])])

    # simulate the pde
    # solve the equation and store the trajectory
    storage = pde.MemoryStorage()
    sol = eq.solve(state, t_range=0.002, dt=1e-5,tracker=storage.tracker(interval=1e-3))
    c_ref = storage.data[2]
```
`simulation_parameters` is a dictionary. I'm filling that one based on reading another file so it doesn't make sense to share that part. But `simulation_parameters['u'][i]` is the velocity in direction i. `simulation_parameters['D'][i]` is the diffusion coefficient of component i. `simulation_parameters['k']` is the reaction rate constant. `simulation_parameters["bc"][i]` is the dirichlet boundary condition of component i at zero (the boundary condition at 1 is a Neumann boundary condition. `simulation_parameters["ini"][i]` is the initial concentration of i
